runner/Dockerfile

```diff
@@ -1,29 +1,40 @@
 FROM ubuntu:18.04
 
 ARG RUNNER_VERSION
-ARG DOCKER_VERSION
+
+ENV container=docker
 
 RUN apt update \
-  && apt install sudo curl ca-certificates -y --no-install-recommends \
-  && curl -L -o docker.tgz https://download.docker.com/linux/static/stable/x86_64/docker-${DOCKER_VERSION}.tgz \
-  && tar zxvf docker.tgz \
-  && install -o root -g root -m 755 docker/docker /usr/local/bin/docker \
-  && rm -rf docker docker.tgz \
-  && curl -L -o /usr/local/bin/dumb-init https://github.com/Yelp/dumb-init/releases/download/v1.2.2/dumb-init_1.2.2_amd64 \
-  && chmod +x /usr/local/bin/dumb-init \
-  && adduser --disabled-password --gecos "" --uid 1000 runner \
-  && usermod -aG sudo runner \
-  && echo "%sudo   ALL=(ALL:ALL) NOPASSWD:ALL" > /etc/sudoers
+  && apt install -y systemd apt-transport-https ca-certificates curl gnupg-agent software-properties-common \
+  && systemctl set-default multi-user.target \
+  && systemctl mask systemd-logind.service systemd-timesyncd.service systemd-resolved.service systemd-networkd.service \
+  && systemctl mask unattended-upgrades.service networkd-dispatcher.service cron.service dev-ttyS0.device \
+  && systemctl disable apt-daily.timer apt-daily-upgrade.timer \
+  && curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add - \
+  && add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable" \
+  && apt update \
+  && apt install -y docker-ce docker-ce-cli containerd.io \
+  && rm -rf /var/lib/apt/lists/*
 
 RUN mkdir -p /runner \
   && cd /runner \
   && curl -L -o runner.tar.gz https://github.com/actions/runner/releases/download/v${RUNNER_VERSION}/actions-runner-linux-x64-${RUNNER_VERSION}.tar.gz \
   && tar xzf ./runner.tar.gz \
   && rm runner.tar.gz \
-  && ./bin/installdependencies.sh
+  && ./bin/installdependencies.sh \
+  && useradd -u 1000 -G sudo,docker -s /sbin/nologin -M runner \
+  && echo "%sudo   ALL=(ALL:ALL) NOPASSWD:ALL" > /etc/sudoers \
+  && chown -R runner:runner /runner
+
+COPY files /tmp/files
+
+RUN cp /tmp/files/entrypoint.sh /usr/local/bin/entrypoint \
+  && cp /tmp/files/actions-runner.sh /usr/local/bin/actions-runner \
+  && cp /tmp/files/actions-runner-halt.sh /usr/local/bin/actions-runner-halt \
+  && cp /tmp/files/actions-runner.service /etc/systemd/system/actions-runner.service \
+  && systemctl enable actions-runner.service \
+  && rm -rf /tmp/files /var/log/* /tmp/* /usr/share/doc* /usr/share/man/* /usr/share/local/*
 
-COPY entrypoint.sh /runner
+STOPSIGNAL 37
```

Comment:
Aha! Neat - I've never realized that the existence of this operation in Dockerfile.

I'm still unsure how this works, though.

I thought that this instructs the container to send SIGALRM(37) to the init process (PID 1 = /sbin/init = systems) on `docker stop`, which might be called via a pod deletion.

But then, who responds to the signal? I "guess" systemd would respond to SIGTERM for stopping. Is it able to respond to 37 as well?

// I read that you've intended to send 37 from within the container itself via the runner unit on post stop using `actions-runner-halt.sh`. Impressive! But I'm still not sure who actually respond to 37 and how.

Reply:
systemd is responsible for this signal in this case.

`STOPSIGNAL 37` indicates the signal to be sent when the container is stopped. The container runtime will use this signal to the container's main process (PID 1, systemd) when it stops the container.

The reason why we use `37 (SIGRTMIN+3)` is based on systemd's container interface document.
https://systemd.io/CONTAINER_INTERFACE/#advanced-integration

> If the container manager wants to cleanly shutdown the container, it might be a good idea to send SIGRTMIN+3 to its init process. systemd will then do a clean shutdown. Note however, that since only systemd understands SIGRTMIN+3 like this, this might confuse other init systems.